include/dali/core/mm/cu_vm.h

```diff
@@ -0,0 +1,166 @@
+    prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+    prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    size_t grain;
+    CUDA_CALL(cuMemGetAllocationGranularity(&grain, &prop, CU_MEM_ALLOC_GRANULARITY_MINIMUM));
+    return grain;
+  };
+  static size_t grain = impl();
+  return grain;
+}
+
+/**
+ * @brief Manages a virtual addresses range
+ */
+class CUMemAddressRange : public UniqueHandle<CUAddressRange, CUMemAddressRange> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(CUAddressRange, CUMemAddressRange);
+
+  static CUMemAddressRange Reserve(size_t size, size_t alignment = 0, CUdeviceptr start = 0) {
+    CUdeviceptr ptr = 0;
+    size_t grain = GetAddressGranularity();
+    size = align_up(size, grain);
+    CUDA_CALL(cuMemAddressReserve(&ptr, size, alignment, start, 0));
+    return CUMemAddressRange(CUAddressRange{ ptr, size });
+  }
+
+  CUdeviceptr ptr() const noexcept {
+      return handle_.ptr();
+  }
+  size_t size() const noexcept {
+      return handle_.size();
+  }
+
+  static void DestroyHandle(CUAddressRange handle) {
+    CUDA_DTOR_CALL(cuMemAddressFree(handle.ptr(), handle.size()));
+  }
+};
+
+/**
+ * @brief Gets default CUmemAllocationProp for allocating memory on given device.
+ * @param device_id device ordinal or -1 for current device.
+ */
+inline CUmemAllocationProp DeviceMemProp(int device_id = -1) {
+  CUmemAllocationProp prop = {};
+  if (device_id < 0) {
+    CUDA_CALL(cudaGetDevice(&device_id));
+  }
+  prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+  prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+  prop.location.id = device_id;
+  return prop;
+}
+
+struct CUMemAllocation : std::pair<CUmemGenericAllocationHandle, size_t> {
+  using std::pair<CUmemGenericAllocationHandle, size_t>::pair;
+  CUmemGenericAllocationHandle &handle() noexcept { return first; }
+  CUmemGenericAllocationHandle  handle() const noexcept { return first; }
+  size_t &size() noexcept { return second; }
+  size_t  size() const noexcept { return second; }
+};
+
+/**
+ * @brief Manages a backing storage block for a virtual address range
+ */
+class CUMem : public UniqueHandle<CUMemAllocation, CUMem> {
+ public:
+  DALI_INHERIT_UNIQUE_HANDLE(CUMemAllocation, CUMem);
+
+  static void DestroyHandle(CUMemAllocation handle) {
+    CUDA_DTOR_CALL(cuMemRelease(handle.first));
+  }
+
+  static CUMem Create(size_t size, int device_id = -1) {
+    return Create(size, DeviceMemProp(device_id));
+  }
+
+  static CUMem Create(size_t size, const CUmemAllocationProp &prop) {
+    size = align_up(size, GetAddressGranularity());
+    CUmemGenericAllocationHandle handle;
+    CUDA_CALL(cuMemCreate(&handle, size, &prop, 0));
+    return CUMem({ handle, size });
+  }
+
+  CUmemGenericAllocationHandle handle() const noexcept { return handle_.first; }
+
+  size_t size() const noexcept { return handle_.second; }
+};
+
+template <typename T = void>
+T *Map(CUdeviceptr virt_addr, CUmemGenericAllocationHandle mem, size_t size, int device_id = -1) {
```

Comment:
As we already have the wrapper types with sizes, do you think a check for that could be in place?

Reply:
We can't check it _here_, can we? This wrapper merely combines cuMemMap with cuMemSetAccess and adds error handling.